rfd/0185-k8s-access-non-cert-routing.md

```diff
@@ -0,0 +1,247 @@
+undertaking this work. For example, using `tsh` to generate a Kubeconfig will
+be faster as it will not need to request the signing of a new certificate as 
+part of the process.
+
+Additionally, from an idealistic design perspective, the current approach
+compromises the intended use of certificates. Certificates are designed to
+encode attributes of an identity - information about which cluster you wish to
+connect to is not an attribute of identity!
+
+## Details
+
+### Today
+
+Today, we encode the intended target cluster within the user X.509 certificate
+within the `KubernetesCluster` attribute (OID 1.9999.1.3).  
+
+This attribute is primarily used by the Proxy and Kubernetes Agent to determine
+where to route requests intended for the Kubernetes cluster.
+
+When the KubernetesCluster attribute is requested by a client using the 
+`GenerateUserCerts` RPC, the certificate request is considered "Kubernetes"
+for analytics purposes. If we were to replace this attribute, we would need to 
+find an alternative way to determine if a certificate request is intended for
+Kubernetes.
+
+The target Kubernetes agent performs authorization checks based on the
+user's identity (e.g roleset, attributes) rather than relying on this attribute.
+
+### Proposal
+
+Rather than leverage the `KubernetesCluster` attribute within the X509
+certificate for routing requests to a specific Kubernetes cluster, we propose to
+pre-pend the request path with this information.
+
+For example, today, to make a request, the client may use the following URL:
+
+```shell
+GET https://example.teleport.sh/apis/apps/v1/namespaces/my-namespace/deployments
+```
+
+Following this proposal, the client would use the following URL:
+
+```shell
+GET https://example.teleport.sh/v1/teleport/base64-teleport-cluster-name/base64-k8s-cluster-name/apis/apps/v1/namespaces/my-namespace/deployments
+```
+
+To avoid potential problems with URL encoding/escaping, the name of the Teleport
+and Kubernetes cluster will be base64 encoded.
+
+Within the Kube Forwarder, new route handlers would be added for all HTTP
+methods for `/v1/teleport/:teleportCluster/:kubeCluster/*`. These route handlers
+would parse the `teleportCluster` and `kubeCluster` from the URL, strip off the
+prefix, and then pass the request through to the original handler - with the
+additional context of the desired Teleport and Kubernetes cluster. This would
+take higher precedence than attributes contained within the certificate.
+
+Teleport clients would no longer need to call `GenerateUserCerts` with a
+`KubernetesCluster` to produce a certificate explicitly for access to
+Kubernetes. When producing a kubeconfig, the client should use the user's 
+standard identity.
+
+To configure Kubernetes clients to work with this, the Kubeconfig generated by
+Teleport would be updated to include this base path within the
+`clusters[name].cluster.server` field:
+
+```yaml
+apiVersion: v1
+clusters:
+- cluster:
```

Comment:
It's important to note that this setup requires a separate cluster entry for each Kubernetes cluster. Currently, our implementation assigns one cluster per Teleport cluster rather than per Kubernetes cluster. This approach has a significant drawback: when two different Kubernetes clusters are part of the same Teleport cluster, `kubectl` caches objects in the same directory `.kube/cache/discovery/teleport_proxy_port`. This can lead to major issues due to version mismatches or incompatible CRDs.

The proposed solution addresses all these problems because kubectl will start caching data in `.kube/cache/discovery/teleport_proxy_port/v1/teleport/base64_teleport_cluster_name/base64_kube_cluster_name`

Check https://github.com/gravitational/teleport/issues/12731